File: ios/BetterBasicStrategy/Models/Hand.swift
import Foundation

enum HandType: String, Codable, Equatable {
    case hard, soft, pair
}

enum HandKey: Codable, Equatable {
    case hard(Int)
    case soft(Int)
    case pair(Int)  // strategy rank (J/Q/K → 10)

    var keyString: String {
        switch self {
        case .hard(let v): return "hard-\(v)"
        case .soft(let v): return "soft-\(v)"
        case .pair(let v): return "pair-\(v)"
        }
    }
}

struct HandState: Codable, Equatable {
    var cards: [Card]
    let total: Int
    let isSoft: Bool
    let isBust: Bool
    let isBlackjack: Bool
    let handType: HandType
    let handKey: HandKey

    static func build(from cards: [Card]) -> HandState {
        let (total, isSoft) = calcTotal(cards)
        let key = getHandKey(cards: cards, total: total, isSoft: isSoft)
        let type_: HandType
        switch key {
        case .pair: type_ = .pair
        case .soft: type_ = .soft
        case .hard: type_ = .hard
        }
        return HandState(
            cards: cards,
            total: total,
            isSoft: isSoft,
            isBust: total > 21,
            isBlackjack: cards.count == 2 && total == 21,
            handType: type_,
            handKey: key
        )
    }
}

func calcTotal(_ cards: [Card]) -> (total: Int, isSoft: Bool) {
    var sum = 0
    var aces = 0
    for card in cards {
        if card.rank == 14 { aces += 1 }
        sum += card.rankValue
    }
    var reduced = 0
    while sum > 21 && reduced < aces {
        sum -= 10
        reduced += 1
    }
    return (sum, reduced < aces)
}

func getHandKey(cards: [Card], total: Int, isSoft: Bool) -> HandKey {
    if cards.count == 2 {
        let r1 = cards[0].strategyRank
        let r2 = cards[1].strategyRank
        if r1 == r2 { return .pair(r1) }
    }
    if isSoft && total >= 13 && total <= 21 { return .soft(total) }
    return .hard(total)
}

File: expo-app/src/engine/hand.ts
import type { Card, HandKey, HandState, HandType } from '../types'
import { rankValue, strategyRank } from './card'

export function calcTotal(cards: Card[]): { total: number; isSoft: boolean } {
  let sum = 0
  let aces = 0
  for (const card of cards) {
    if (card.rank === 14) aces++
    sum += rankValue(card.rank)
  }
  let reduced = 0
  while (sum > 21 && reduced < aces) { sum -= 10; reduced++ }
  return { total: sum, isSoft: reduced < aces }
}

export function getHandKey(cards: Card[]): HandKey {
  const { total, isSoft } = calcTotal(cards)
  if (cards.length === 2) {
    const r1 = strategyRank(cards[0].rank)
    const r2 = strategyRank(cards[1].rank)
    if (r1 === r2) return { type: 'pair', rank: r1 }
  }
  if (isSoft && total >= 13 && total <= 21) return { type: 'soft', total }
  return { type: 'hard', total }
}

export function getHandType(cards: Card[]): HandType {
  return getHandKey(cards).type
}

export function buildHandState(cards: Card[]): HandState {
  const { total, isSoft } = calcTotal(cards)
  return {
    cards,
    total,
    isSoft,
    isBust: total > 21,
    isBlackjack: cards.length === 2 && total === 21,
    handType: getHandType(cards),
    handKey: getHandKey(cards),
  }
}
